### PR TITLE
Chatglm2 rope optimization on xpu

### PR DESCRIPTION
**Pull request by qiuxin2012**
## Description

Speed up chat glm2 on xpu

### 1. Why the change?

Speed up chat glm2 on xpu with ipex

### 2. Summary of the change 
1. use `torch.ops.torch_ipex.apply_rotary_embedding` instead of apply_rotary_pos_emb
2. (sin, cos)'s repeat_interleave in model's forward if xpu is used. (cpu unchanged)

### CODE REVIEW

File: python/llm/src/bigdl/llm/transformers/models/chatglm2.py [hunk 1]
```diff
@@ -87,6 +88,75 @@ def chatglm_rms_norm_forward(self, hidden_states):
...
+    if position_ids is not None:
+        rotary_pos_emb = rotary_pos_emb[position_ids]
+    else:
+        rotary_pos_emb = rotary_pos_emb[None, :seq_length]
+    if use_fuse_rope:
+        # repeat cos sin here.
+        cos, sin = rotary_pos_emb.split(rotary_pos_emb.shape[-1] // 2, dim=-1)
+        cos = cos.squeeze(-1)
+        sin = sin.squeeze(-1)
+        cos = torch.repeat_interleave(cos[:, :, None, :], 2, 3)
```
Comment (hkvision) [inline python/llm/src/bigdl/llm/transformers/models/chatglm2.py:136]: Need to add a comment why changes here

Comment (qiuxin2012) [reply python/llm/src/bigdl/llm/transformers/models/chatglm2.py:138]: Added.

File: python/llm/src/bigdl/llm/transformers/models/chatglm2.py [hunk 2]
```diff
@@ -87,6 +88,77 @@ def chatglm_rms_norm_forward(self, hidden_states):
...
+
+    if full_attention_mask is None:
+        if (attention_mask is not None and not attention_mask.all()) or (
+                past_key_values and seq_length != 1):
+            full_attention_mask = self.get_masks(input_ids,
+                                                 past_key_values,
+                                                 padding_mask=attention_mask)
+
+    use_fuse_rope = input_ids.device.type == "xpu"
+    use_fuse_rope = use_fuse_rope and not self.training
```
Comment (hkvision) [inline python/llm/src/bigdl/llm/transformers/models/chatglm2.py:123]: we can combine this to `if input_ids.device.type == "xpu" and not self.training`

**Review by hkvision — APPROVED** [review]

**Comment by qiuxin2012** [comment]
Upstream output:
```
Once upon a time, there was a young girl named Samantha who lived with her parents in a small town. Samantha had always dreamed of traveling the world and experiencing new cultures. One day, she heard about a travel agency that offered guided tours to far-off lands.

The travel agency offered a variety

```
PR's output:
```
Once upon a time, there was a young girl named Samantha who lived with her parents in a small town. Samantha had always dreamed of traveling the world and experiencing new cultures. One day, she heard about a travel agency that offered guided tours to far-off lands.

The travel agency offered a variety
```

**Comment by jason-dai** [comment]
Need to add test similar to https://github.com/intel-analytics/BigDL/pull/9347

**Comment by qiuxin2012** [comment]
> Need to add test similar to #9347

Correctness test in #9450

**Pull request merged**